New constant for breaking changes in libcmark 0.29.0

anthonyryan1 opened the pull request:
Safe is now enabled by default and a no-op, which will probably
come as a surprise to some people upgrading. People wanting the
old default behavior will need to use the new Unsafe constant.

I'm not sure if we'll want to deprecate the Safe constant over
a couple years or just leave it since it's a no-op.

dwo0 commented:
There are three other constants that have been added to libcmark besides Unsafe: `CMARK_OPT_NORMALIZE`, `CMARK_OPT_VALIDATE_UTF8` and `CMARK_OPT_SMART`. Is there any value in including these constants as well?

krakjoe commented:
@anthonyryan1 thanks ... I think we ought to treat constants the same as upstream, so a no-op is fine.

@dwo0 adding those seems reasonable, right ?

remicollet commented:
Does raising minimal supported library version really make sense ?

krakjoe commented:
@remicollet I'm not sure actually, did any distros bump, what do you think ?

remicollet commented:
@krakjoe Fedora have 0.28.3
https://rpms.remirepo.net/rpmphp/zoom.php?rpm=cmark

krakjoe commented:
@anthonyryan1 I think the only thing we can reasonably do here is register the constant for versions that support it, and carry the breaking change ... it's awkward, but < 1.0, so kinda expected ...

anthonyryan1 committed:
Pass CMARK_OPT_UNSAFE through to PHP if it's defined

Also add the other missing options in case anyone wants to use them in the future.

anthonyryan1 commented:
One year later.

I've updated this pull request based on the feedback above.

* I have added the other missing render options.
* Instead of bumping the library requirement, we're going to pass through the new constant if it's exposed (aka, built against libcmark 0.29.0)

Feedback welcome!

dwo0 commented:
I like it.

dwo0 commented:
@anthonyryan1, actually, you may have an error on line 154. `CMARK_OPT_SMART` should be `CMARK_OPT_VALIDATE_UTF8`.

dwo0 commented on src/render.c:
```diff
@@ -146,6 +146,12 @@ PHP_MINIT_FUNCTION(CommonMark_Render)
+#endif
+	REGISTER_NS_LONG_CONSTANT("CommonMark\\Render", "Normalize", CMARK_OPT_NORMALIZE, CONST_CS|CONST_PERSISTENT);
+	REGISTER_NS_LONG_CONSTANT("CommonMark\\Render", "Smart", CMARK_OPT_SMART, CONST_CS|CONST_PERSISTENT);
+	REGISTER_NS_LONG_CONSTANT("CommonMark\\Render", "ValidateUTF8", CMARK_OPT_SMART, CONST_CS|CONST_PERSISTENT);
```
`REGISTER_NS_LONG_CONSTANT("CommonMark\\Render", "ValidateUTF8", CMARK_OPT_VALIDATE_UTF8, CONST_CS|CONST_PERSISTENT);`

anthonyryan1 replied:
Good catch, corrected.
